### DISCUSSION TITLE
2021-06-23 retrospective & plan discussion

### DISCUSSION THREAD

**Discussion by kgodey** [mathesar-foundation/mathesar#278] [Maintainer discussions]
# Retrospective
This covers work since our last meeting on 2021-06-15.

## Product
_No product work this week._

## Design
- Design completed and review started for [4. Editable Tables milestone](https://github.com/centerofci/mathesar/milestone/4):
    - Add, Edit and Delete Records
- Finalized design for [Lookup Component Specs](https://github.com/centerofci/mathesar/discussions/233)

## Development
- Backend work for syncing state between the database and web application has been completed and merged.
- Backend work on [2. Tables for File Import milestone](https://github.com/centerofci/mathesar/milestone/1):
    - Table column type inference
    - DB function to edit column names
    - Additional filters for table list endpoint
    - Finish TSV file upload
    - Duplicate schema bug fixed
    - Column API spec is ongoing
- Frontend work on [1. Viewing Tables milestone](https://github.com/centerofci/mathesar/milestone/3)
    - Table pagination, url sync
        - Button component
        - Dropdown component
        - Basic select component
        - Skeleton loader
        - Portal action
        - Popper action
        - Click off bound callback action

## Infrastructure
- Work on setting up the staging environment is ongoing.

# Active Work
This is work planned for the upcoming week.

General notes:
- Pavish is full-time now (yay!), so frontend work will speed up.
- This is our first week working completely asynchronously.
- We're doing our first team event this week!

## Product
_No product work this week._

## Design
- @ghislaineguerin: Finalize design for [4. Editable Tables milestone](https://github.com/centerofci/mathesar/milestone/4):
    - Add, Edit and Delete Records
- @ghislaineguerin: Design for [5\. Multiple Database Support](https://github.com/centerofci/mathesar/issues/172):
    - Design for switching between multiple databases

## Development
- Backend work on [2. Tables for File Import milestone](https://github.com/centerofci/mathesar/milestone/1):
    - @eito-fis: Finish table type suggestion endpoint
    - @mathemancer: Finish Column API spec and begin implementation
    - @mathemancer: Complete other API work blocking tables for file import 
- Backend work on [3. Basic Filtering, Sorting, Grouping](https://github.com/centerofci/mathesar/milestone/2)
    - @eito-fis: Sorting and filtering the record list endpoint
- Frontend work on [1. Viewing Tables milestone](https://github.com/centerofci/mathesar/milestone/3)
    - @pavish: Sidebar: Search through tables/schemas
    - @pavish: ARIA implementation for components
    - @pavish: Start on unit tests (Not sure if it will be complete this week)

## Infrastructure
- @kgodey: Complete setting up staging server
- @kgodey: If time permits, start setting up API documentation

# Questions for Discussion
- Milestone [1. Viewing Tables](https://github.com/centerofci/mathesar/milestone/3) is due this week. Is anything blocking it from being done?
- Does everyone have enough to work on this week?
- How's working asynchronously going?
- Any questions or comments on any of the above?

**Comment by pavish** [comment]
> Milestone 1. Viewing Tables is due this week. Is anything blocking it from being done?

There's nothing blocking it on the frontend. The functional aspects for the milestone will be done, but not sure if we'll get enough time to get the unit tests and ARIA accessibility for all the components done. (Since these components are common, I think we can spread this out across milestones, as they don't affect the functionality)

> Does everyone have enough to work on this week?

Yes. I do.

> How's working asynchronously going?

It's going good, I do not see any challenges on my end.

> Any questions or comments on any of the above?

None.

**Reply by kgodey** [reply]
@pavish sounds good, please make issues for the unit tests and ARIA accessibility issues that remain after the first milestone and I'll put them in other milestones.

**Comment by kgodey** [comment]
My answers:

> Milestone 1. Viewing Tables is due this week. Is anything blocking it from being done?

I don't think so but @pavish please double check the the search endpoints for tables and schemas work as expected so we can make any fixes if needed ASAP.

> Does everyone have enough to work on this week?

I do, the infrastructure work is taking me longer than anticipated. If I get done with it early, I'll pick a new backend task from the pool.

> How's working asynchronously going?

It's going very well for me, I can focus better, and the impromptu meetings I've had have had a clearer agenda. 

> Any questions or comments on any of the above?

None.

**Comment by mathemancer** [comment]
My answers:

> Milestone 1. Viewing Tables is due this week. Is anything blocking it from being done?

Nothing to my knowledge

> Does everyone have enough to work on this week?

I do, I'm a little slow this week while my eyes recover.  

> How's working asynchronously going?

I really like it so far.  Comms feel more intentional.

> Any questions or comments on any of the above?

None.

BONUS Answer to an unasked question:

I'd like to flag #280 since it's going to get worse quickly if we don't handle it.

Edit:  @kgodey noticed #280 while I was composing, and already put it in the next milestone.  Good show!

**Reply by eito-fis** [reply]
Re: #280, do we want to open a broader issue on the topic of engine overuse? I meant for that issue to be specific to the caching problem, and I imagine there are other causes for the overall issue.

**Reply by kgodey** [reply]
@eito-fis could you describe the broader problem of engine overuse? I was thinking that we'd just cache an engine per database and use the cached engine everywhere, but I'm not sure that I'm thinking through the problem correctly.

**Reply by eito-fis** [reply]
@kgodey: I was mostly asking because @mathemancer brought up in our meeting that we haven't been very consistent in our use of engines so far, and there might be other places were we make too many engines (one obvious case is the use of `create_mathesar_engine()` in our CSV import functions).

**Reply by kgodey** [reply]
Makes sense. I'll leave it up to @mathemancer whether we want to broaden the scope of #280 or create a new issue.

**Reply by mathemancer** [reply]
I'll ticket that work as soon as I get time.  I think the order should be to first get some engines _created_ that are globally available, and that are associated with the underlying server process.  I think we'll want some kind of function that "gets the engine", and either creates it, or gets it if it's already extant.  (I.e., instead of `create_mathesar_engine`, we'll just call `get_mathesar_engine` in every such place).  

After we have that function set up, we can just drop it in anywhere `create_mathesar_engine` is currently used.

The function should also have a recovery ability, since it may be that the engine ends up broken after being up for 3 days or whatever.  I.e., the function should be able to check that the engine works and recreate it if necessary.

**Comment by eito-fis** [comment]
My answers:
>  Milestone 1. Viewing Tables is due this week. Is anything blocking it from being done?

Nothing that I know of.

> Does everyone have enough to work on this week?

Still have the record list filtering work on the backlog, so plenty to do.

> How's working asynchronously going?

Overall well. I like having long stretches of time to work, but I do feel like it take a bit longer for me to get un-stuck on some things.

> Any questions or comments on any of the above?

N/A

**Comment by ghislaineguerin** [comment]
> Milestone 1. Viewing Tables is due this week. Is anything blocking it from being done?

Nothing that I know of.

> Does everyone have enough to work on this week?

Yes. I'm also adding a new issue #285 to work on this week.

> How's working asynchronously going?

It's going well. I like to be able to jump on calls to resolve blockers. I liked how discussions are being used, for example in the case of design specs improvements.

> Any questions or comments on any of the above?

None.
